packages/context/src/manifest.ts

```diff
@@ -0,0 +1,227 @@
+
+  if (dataSources && serviceName !== undefined) {
+    const defaultModelDataSource = dataSources[serviceName];
+
+    return defaultModelDataSource?.uri;
+  }
+  return undefined;
+}
+
+function getFlexEnabled(manifest: Manifest): boolean {
+  return manifest["sap.ui5"]?.flexEnabled ?? false;
+}
+function getMinUI5Version(manifest: Manifest): string | undefined {
+  return manifest["sap.ui5"]?.dependencies?.minUI5Version;
+}
+
+/**
+ * Extract details information from manifest
+ * @param manifest manifest of an app
+ */
+async function extractManifestDetails(
+  manifest: Manifest
+): Promise<ManifestDetails> {
+  const customViews = {};
+  const targets = manifest["sap.ui5"]?.routing?.targets;
+  if (targets) {
+    for (const name of Object.keys(targets)) {
+      const target = targets[name];
+      if (target) {
+        const settings = (target?.options as any)?.settings;
+        if (settings?.entitySet && settings?.viewName) {
+          customViews[settings.viewName] = {
+            entitySet: settings.entitySet,
+          };
+        }
+        if (settings?.entitySet && settings?.content) {
+          // search for custom section and get its entity set
+          const templateKey =
+            settings?.content?.body?.sections?.CustomSection.template;
+          if (templateKey) {
+            customViews[templateKey] = {
+              entitySet: settings?.entitySet,
+            };
+          }
+        }
+      }
+    }
+  }
+
+  const mainServiceName = getMainService(manifest) ?? "";
+  const mainServicePath = getServicePath(manifest, mainServiceName) ?? "/";
+  const flexEnabled = getFlexEnabled(manifest);
+  const minUI5Version = getMinUI5Version(manifest);
+  return {
+    mainServicePath,
+    customViews,
+    flexEnabled,
+    minUI5Version,
+  };
+}
+
+/**
+ * Get details of manifest defined under `webapp`
+ * @param documentPath path to a file i.e absolute/path/webapp/ext/main/Main.view.xml
+ */
+export async function getManifestDetails(
+  documentPath: string
+): Promise<ManifestDetails> {
+  const manifestPath = await findManifestPath(documentPath);
+  if (!manifestPath) {
+    return {
+      flexEnabled: false,
+      customViews: {},
+      mainServicePath: undefined,
+      minUI5Version: undefined,
+    };
+  }
+  const manifest = await getUI5Manifest(manifestPath);
+  if (!manifest) {
+    return {
+      flexEnabled: false,
+      customViews: {},
+      mainServicePath: undefined,
+      minUI5Version: undefined,
+    };
+  }
+  return extractManifestDetails(manifest);
+}
+
+/**
+ * Get custom view id
+ *
+ * Id of a `.view.xml` or `.fragment.xml` is calculated based on manifest app id and relative file from app root folder
+ */
+export const getCustomViewId = async (
+  documentPath: string
+): Promise<string> => {
+  const appRoot = await findAppRoot(documentPath);
+  if (!appRoot) {
+    return "";
+  }
+  const manifestPath = (await findManifestPath(documentPath)) as string;
```

Comment:
Why do we need to cast here? We should avoid casting, because this can lead to runtime errors.